### PR TITLE
Change depth to MediaTrackConstraints

### LINKED ISSUE #143: Change depth to MediaTrackConstraints
Current spec defines the depth as a new media type in `MediaStreamConstraints`

```
partial dictionary MediaStreamConstraints {
    (boolean or MediaTrackConstraints) depth = false;
};
```

However actually the depth camera is a kind of video capture device, it shares most of characteristics of a color camera. The major difference with a color camera is the pixel format and the meta data describing the pixel value (depth value).

So I propose to change the depth to the `MediaTrackConstraints`. To accommodate more kinds of video capture device, it would introduce the `videoKind` constrainable property:

```
partial dictionary MediaTrackConstraintSet {
             ConstrainDOMString videoKind;
};

enum VideoDeviceKindEnum {
    "color",
    "depth",
    "infrared",
    "fisheye"
};
```

The example of requesting a single depth track:
```
navigator.mediaDevices.getUserMedia({
  video: {videoKind: {exact: "depth"}}
}).then(function (stream) {
  // Render the depth video.
  var video = document.querySelector('#depth-video');
  video.srcObject = stream;
  video.play()
});
```

The example of requesting the color and depth tracks from one 3D camera by setting the `groupId` of the 3D camera.
```
navigator.mediaDevices.getUserMedia({
  video: {groupId: {exact: id}}
}).then(function (stream) {
  // Render the color video.
  var video = document.querySelector('#color-video');
  video.srcObject = stream;
  video.play()
});

navigator.mediaDevices.getUserMedia({
  video: {videoKind: {exact: "depth"}, groupId: {exact: id}}
}).then(function (stream) {
  // Render the depth video.
  var video = document.querySelector('#depth-video');
  video.srcObject = stream;
  video.play()
});
```

@robman @anssiko @fluffy @astojilj,  WDYT?

### PR DESCRIPTION
**Pull request by huningxin**
Fix #143

### CODE REVIEW

**Comment by huningxin** [comment]
@anssiko @robman @astojilj, PTAL.

BTW, I don't quite get the respec errors in the travis-ci:
```
😱 ReSpec error: Multiple <dfn>s for videoKind in MediaTrackConstraintSet
```

I actually only define `videoKind` once in `MediaTrackConstraintSet`. Did I miss anything?

**Comment by anssiko** [comment]
Thanks for the PR!

It seems the remaining CI errors are false negatives due to a ReSpec limitation: ReSpec complains if the same dictionary member names are used and defined across multiple dictionaries. If we'd leave the `<dfn>`s out from other dictionaries, I believe ReSpec would give an error "`No <dfn> for XXXXX`".

**Comment by anssiko** [comment]
I just validated my hypothesis: if we drop the other `<dfn>`s for `videoKind` we'll get the following warnings:
```
No <dfn> for videoKind in MediaTrackCapabilities.
No <dfn> for videoKind in MediaTrackConstraintSet.
No <dfn> for videoKind in MediaTrackSettings.
```

The same applies to the following members too defined in multiple dictionaries (the CI check halts on the first error so these go unnoticed):
```
depthNear
depthFar
focalLengthX
focalLengthY
```

Given "`No <dfn> for XXXXX`" warnings is better than the current errors, I suggest we keep `<dfn>`s for the above dictionary members only in `MediaTrackSupportedConstraints` and remove `<dfn>`s from any other dictionaries as we figure out how to fix this properly.

That said, the warnings would not block this PR from being merged.

**Comment by anssiko** [comment]
@huningxin I figured this out. The proper fix for this is to use the `data-dfn-for` attribute as follows: `<dfn data-dfn-for="MediaTrackCapabilities"><code>videoKind</code></dfn>` and the same for the other dictionaries and members.

See the ReSpec docs: https://github.com/w3c/respec/wiki/data-dfn-for

Could you update this PR accordingly?

File: index.html [hunk 1]
```diff
@@ -683,115 +681,396 @@
       </h2>
       <section>
         <h2>
-          <code><a>MediaStreamConstraints</a></code> dictionary
```
Comment (anssiko) [inline index.html:686]: We no longer extend `MediaStreamConstraints` so we can drop that from the Dependencies section.

**Comment by anssiko** [comment]
@huningxin When you've done the updates, could you push the updated `index.src.html` too?

To build `index.html` from `index.src.html`, just `$ node build.js`. (We should probably document this workflow somewhere :-) - edit: documented the workflow at https://github.com/w3c/mediacapture-depth#pull-requests)

File: index.html (hunk 1 above)
Comment (huningxin) [reply index.html:686]: Fixed.

**Comment by huningxin** [comment]
> @huningxin I figured this out. The proper fix for this is to use the data-dfn-for attribute as follows: <dfn data-dfn-for="MediaTrackCapabilities"><code>videoKind</code></dfn> and the same for the other dictionaries and members.

I tried `data-dfn-for`. It doesn't solve the errors. I actually think it is an respec bug, as I already define the `data-dfn-for` in the 
```
<dl data-link-for="MediaTrackSupportedConstraints" data-dfn-for=
          "MediaTrackSupportedConstraints" class="dictionary-members">
```

It follows the mediacapture-main: https://github.com/w3c/mediacapture-main/blob/master/getusermedia.html#L1336

**Comment by huningxin** [comment]
> @huningxin When you've done the updates, could you push the updated index.src.html too?
> 
> To build index.html from index.src.html, just $ node build.js. (We should probably document this workflow somewhere :-) - edit: documented the workflow at https://github.com/w3c/mediacapture-depth#pull-requests)

@anssiko , thanks for the instructions. I updated the index.src.html and build to index.html.

Please take another look.

**Comment by huningxin** [comment]
The warnings:
```
line 503 column 89 - Warning: trimming empty <span>
line 504 column 42 - Warning: trimming empty <span>
line 605 column 92 - Warning: trimming empty <span>
line 606 column 48 - Warning: trimming empty <span>
```

It is introduced by MathJax generation code.

**Comment by anssiko** [comment]
@huningxin Here's a PR to fix all the ReSpec errors: https://github.com/huningxin/mediacapture-depth/pull/1

LGTM otherwise. I'll merge this PR when you land the above PR, unless someone else has comments.

(We don't need to worry about the MathJax warnings.)

**Comment by huningxin** [comment]
> @huningxin Here's a PR to fix all the ReSpec errors: huningxin#1

@anssiko , thanks much for the fix! It is merged.

**Comment by anssiko** [comment]
Merged, and the [Editor's Draft][1] is updated accordingly.

@huningxin Thanks for fixing this major issue!

[1]: https://w3c.github.io/mediacapture-depth/

**Pull request merged**